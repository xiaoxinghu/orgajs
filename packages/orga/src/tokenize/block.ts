import { Reader } from '../reader'
import { BlockBegin, BlockEnd } from '../types'

interface Props {
  reader: Reader;
}

export default ({ reader }: Props): [BlockBegin | BlockEnd, () => void] | undefined => {
  const { match, eat } = reader

  let m = match(/^\s*#\+begin_([^\s]+)\s*(.*)$/i)
  if (m) {
    const params = m.captures[2].split(' ').map(p => p.trim()).filter(String)
    return [{
      type: 'block.begin',
      name: m.captures[1],
      params,
      position: m.position,
    }, () => eat('line')];
  }

  m = match(/^\s*#\+end_([^\s]+)\s*$/i)
  if (m) {
    return [{
      type: 'block.end',
      position: m.position,
      name: m.captures[1],
    }, () => eat('line')];
  }
}
