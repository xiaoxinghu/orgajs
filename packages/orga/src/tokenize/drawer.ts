import { Reader } from '../reader'
import { Token } from '../types'
import { tokDrawerBegin, tokDrawerEnd } from './util';

interface Props {
  reader: Reader;
}

export default ({ reader }: Props): Token[] => {
  const { match, eat } = reader

  const m = match(/^:(\w+):(?=\s*$)/)
  if (m) {
    eat('line')
    const name = m.captures[1]
    const position = m.position;
    if (name.toLowerCase() === 'end') {
      return [tokDrawerEnd({ position })];
    } else {
      return [tokDrawerBegin(name, { position })];
    }
  }

  return []
}
