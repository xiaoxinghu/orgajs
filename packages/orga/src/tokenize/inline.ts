import { Point } from 'unist'
import { isEqual } from '../position'
import { Reader } from '../reader'
import { FootnoteReference, Link, StyledText, Token, Newline } from './types'
import uri from '../uri'
import { escape } from '../utils'
import * as tk from './util';

const POST = `[\\s-\\.,:!?'\\)}]|$`
const BORDER = `[^,'"\\s]`

const MARKERS: { [key: string]: StyledText['type'] } = {
  '*': 'text.bold',
  '=': 'text.verbatim',
  '/': 'text.italic',
  '+': 'text.strikeThrough',
  '_': 'text.underline',
  '~': 'text.code',
}

interface Props {
  reader: Reader
  start?: Point
  end?: Point
}

export const tokenize = ({ reader, start, end }: Props, { ignoring }: { ignoring: string[] } = { ignoring: [] }): Token[] => {
  const { now, eat, eol, match, jump, substring, getChar } = reader
  start = start || { ...now() }
  end = end || { ...eol() }
  jump(start)

  let cursor: Point = { ...start }

  const _tokens: Token[] = []

  const tokLink = (): Link => {
    const m = match(/^\[\[([^\]]*)\](?:\[([^\]]*)\])?\]/m)
    if (!m) return undefined
    const linkInfo = uri(m.captures[1])
    const { value, ...rest } = linkInfo;
    return tk.tokLink(value, { ...rest, description: m.captures[2], position: m.position });
  }

  const tokFootnoteAnonOrInline = (): Token[] => {
    const tokens: Token[] = [];

    let m = match(/^\[fn:(\w*):/);
    if (!m) return [];
    tokens.push(tk.tokFootnoteInlineBegin(m.captures[1], { position: m.position }));
    jump(m.position.end);

    m = match(/^\]/);
    if (m) {
      // empty body
      tokens.push(tk.tokText('', { position: { start: m.position.start, end: m.position.start, indent: m.position.indent } }));
    } else {
      tokens.push(...tokenize({ reader }, { ignoring: [']'] }));
    }

    m = match(/^\]/);
    if (!m) return [];
    tokens.push(tk.tokFootnoteReferenceEnd({ position: m.position }));

    jump(tokens[0].position.start);

    return tokens;
  }

  const tokFootnote = (): FootnoteReference => {
    const m = match(/^\[fn:(\w+)\]/);
    if (m) {
      return tk.tokFootnoteReference(m.captures[1], { position: m.position });
    }
  }

  const tokStyledText = (marker: string) => (): StyledText => {
    const m = match(
      RegExp(`^${escape(marker)}(${BORDER}(?:.*?(?:${BORDER}))??)${escape(marker)}(?=(${POST}.*))`, 'm'))
    if (!m) return undefined
    return {
      type: MARKERS[marker],
      value: m.captures[1],
      position: m.position,
    }
  }

  const tryToTokens = (tok: () => Token[]) => {
    const tokens = tok()
    if (tokens.length === 0) return false
    cleanup()
    _tokens.push(...tokens)
    jump(tokens[tokens.length - 1].position.end)
    cursor = { ...now() }
    return true
  }

  const tryTo = (tok: () => Token) => {
    return tryToTokens(() => {
      const r = tok();
      return r ? [r] : [];
    });
  }

  const cleanup = () => {
    if (isEqual(cursor, now())) return
    const position = { start: { ...cursor }, end: { ...now() } }
    const value = substring(position)
    _tokens.push(tk.tokText(value, { position: position }));
  }

  const tokNewline = (): Newline => {
    const m = match(/^\n/)
    if (!m) return undefined
    return tk.tokNewline({ position: m.position });
  }

  const tok = (): Token[] => {
    if (isEqual(now(), end)) {
      return
    }
    const char = getChar()

    if (ignoring.includes(char)) {
      return [];
    }

    if (char === '[') {
      if (tryTo(tokLink)) return tok()
      if (tryTo(tokFootnote)) return tok()
      if (tryToTokens(tokFootnoteAnonOrInline)) return tok();
    }

    if (MARKERS[char]) {
      const pre = getChar(-1)
      if (now().column === 1 || /[\s({'"]/.test(pre)) {
        if (tryTo(tokStyledText(char))) return tok()
      }
    }

    if (tryTo(tokNewline)) return tok()

    eat()
    tok()
  }

  tok()
  cleanup()
  return _tokens

}
