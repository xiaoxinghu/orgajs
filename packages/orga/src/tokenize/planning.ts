import { Point } from 'unist'
import { Reader } from '../reader'
// import { inspect }  from 'util'
import { parse as parseTimestamp } from '../timestamp'
import { Token } from '../types'
import * as tk from './util';

interface Props {
  reader: Reader;
  keywords: string[];
  timezone: string;
}


export default ({ reader, keywords, timezone }: Props): Token[] => {
  const { eat, substring, now, getLine } = reader

  const p = RegExp(`(${keywords.join('|')}):`, 'g')

  const currentLine = getLine()

  const { line, column } = now()

  const getLocation = (offset: number): Point => ({
    line, column: column + offset,
  })

  const all: Token[] = []

  const parseLastTimestamp = (end: number) => {
    if (all.length === 0) return
    const { type, position } = all[all.length - 1]
    if (!position) throw Error(`position is ${position}`)
    if (type !== 'planning.keyword') return
    const endLocation = getLocation(end)
    const timestampPosition = { start: position.end, end: endLocation }
    const value = substring(timestampPosition)
    const stamp = parseTimestamp(value, { timezone });
    if (stamp) {
      all.push(tk.tokPlanningTimestamp(stamp, {
        position: timestampPosition
      }));
    }
  }

  let m
  while ((m = p.exec(currentLine)) !== null) {
    parseLastTimestamp(m.index)

    all.push(tk.tokPlanningKeyword(m[1], {
      position: {
        start: getLocation(m.index),
        end: getLocation(p.lastIndex),
      }
    }));
  }
  parseLastTimestamp(currentLine.length)
  eat('line')

  // console.log(inspect(all, false, null, true))
  return all
}
