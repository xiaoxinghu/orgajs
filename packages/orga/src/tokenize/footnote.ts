import { Reader } from '../reader'
import { Token } from '../types'
import { tokenize as tokenizeInline } from './inline'

interface Props {
  reader: Reader;
}

export default ({ reader }: Props) : Token[] => {
  const { match, jump, eat } = reader
  let tokens: Token[] = []
  const m = match(/^\[fn:([^\]]+)\]/)
  if (!m) return []
  tokens.push({
    type: 'footnote.label',
    label: m.captures[1],
    position: m.position,
  })
  jump(m.position.end)
  eat('whitespaces')

  tokens = tokens.concat(tokenizeInline({ reader }))

  return tokens
}
