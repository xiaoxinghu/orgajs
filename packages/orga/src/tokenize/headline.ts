import { isEmpty } from '../position'
import { Reader } from '../reader'
import { TodoKeywordSet } from '../todo-keyword-set'
import { Token } from '../types'
import { tokenize } from './inline'

interface Props {
  reader: Reader;
  todoKeywordSets: TodoKeywordSet[];
}

export default ({ reader, todoKeywordSets }: Props) : Token[] => {

  const {
    match,
    now,
    eol,
    eat,
    jump,
    substring,
  } = reader

  // TODO: cache this, for performance sake
  const todos = todoKeywordSets.flatMap(s => s.keywords)

  const isActionable = (keyword: string): boolean => {
    return !!todoKeywordSets.find(s => s.actionables.includes(keyword))
  }

  let buffer: Token[] = []

  const stars = eat(/^\*+(?=\s)/)
  if (isEmpty(stars.position)) throw Error('not gonna happen')
  buffer.push({
    type: 'stars',
    level: stars.value.length,
    position: stars.position,
  })

  eat('whitespaces')
  const keyword = eat(RegExp(`^${todos.map(escape).join('|')}`))
  if (!isEmpty(keyword.position)) {
    buffer.push({
      type: 'todo',
      keyword: keyword.value,
      actionable: isActionable(keyword.value),
      position: keyword.position,
    })
  }
  eat('whitespaces')
  const priority = eat(/^\[#(A|B|C)\]/)
  if (!isEmpty(priority.position)) {
    buffer.push({
      type: 'priority',
      ...priority,
    })
  }

  eat('whitespaces')

  const tags = match(/\s+(:(?:[\w@_#%-]+:)+)[ \t]*$/gm)
  let contentEnd = eol()
  if (tags) {
    contentEnd = tags.position.start
  }

  const tokens = tokenize({ reader, end: contentEnd })

  buffer = buffer.concat(tokens)


  if (tags) {
    eat('whitespaces')
    const tagsPosition = { start: now(), end: tags.position.end }
    const s = substring(tagsPosition)
    buffer.push({
      type: 'tags',
      tags: s.split(':').map(t => t.trim()).filter(Boolean),
      position: { start: now(), end: tags.position.end },
    })
    jump(tags.position.end)
  }
  eat('line')
  return buffer
}
