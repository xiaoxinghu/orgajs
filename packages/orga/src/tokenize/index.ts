import defaultOptions, { ParseOptions } from '../options'
import { isEmpty } from '../position'
import { read } from '../reader'
import todoKeywordSet, { TodoKeywordSet } from '../todo-keyword-set'
import { Token } from '../types'
import tokenizeBlock from './block'
import tokenizeDrawer from './drawer'
import tokenizeFootnote from './footnote'
import tokenizeHeadline from './headline'
import { tokenize as inlineTok } from './inline'
import tokenizeListItem from './list'
import tokenizePlanning from './planning'
import tokenizeTable from './table'
import { Position } from 'unist'

const PLANNING_KEYWORDS = ['DEADLINE', 'SCHEDULED', 'CLOSED']

export interface Lexer {
  eat: (type?: string) => Token | undefined;
  eatAll: (type: string) => number;
  peek: (offset?: number) => Token | undefined;
  match: (cond: RegExp | string, offset?: number) => boolean;
  all: () => Token[];
  save: () => number;
  restore: (point: number) => void;
  addInBufferTodoKeywords: (text: string) => void;
  substring: (position: Position) => string;
  /** Modify the next token (or the token at the given offset). */
  modify(f: (t: Token) => Token, offset?: number): void;
}

export const tokenize = (text: string, options: Partial<ParseOptions> = {}): Lexer => {

  const { timezone, todos } = { ...defaultOptions, ...options }

  const reader = read(text)

  const state: {
    context: 'verse' | null;
  } = {
    context: undefined
  };

  const setContext = (context: 'verse') => {
    state.context = context;
  }

  const removeContext = (matchContext: NonNullable<(typeof state)['context']>) => {
    if (state.context === matchContext) {
      state.context = null;
    }
  }

  const {
    isStartOfLine,
    eat,
    getLine,
    getChar,
    now,
    match,
    EOF,
    substring,
  } = reader

  const globalTodoKeywordSets = todos.map(todoKeywordSet)

  const inBufferTodoKeywordSets: TodoKeywordSet[] = []

  const todoKeywordSets = () => {
    return inBufferTodoKeywordSets.length === 0
      ? globalTodoKeywordSets : inBufferTodoKeywordSets
  }

  let tokens: Token[] = []

  let cursor = 0

  const tok = (): Token[] => {
    eat('whitespaces')
    if (EOF()) return []

    if (getChar() === '\n') {
      return [{
        type: 'newline',
        position: eat('char').position,
      }]
    }

    if (isStartOfLine() && match(/^\*+\s+/) && state.context !== 'verse') {
      return tokenizeHeadline({
        reader,
        todoKeywordSets: todoKeywordSets(),
      })
    }

    const l = getLine()
    if (PLANNING_KEYWORDS.some((k) => l.startsWith(k))) {
      return tokenizePlanning({
        reader,
        keywords: PLANNING_KEYWORDS,
        timezone })
    }

    if (l.startsWith('#+')) {

      const keyword = match(/^\s*#\+(\w+):\s*(.*)$/)
      if (keyword) {
        eat('line')
        return [{
          type: 'keyword',
          key: keyword.captures[1],
          value: keyword.captures[2],
          position: keyword.position,
        }]
      }

      const tokBlock = tokenizeBlock({ reader })
      if (tokBlock) {
        const [block, action] = tokBlock;
        // when in a verse block, the only block begin/end that we
        // accept is a verse end - all others are parsed as text
        if (!(state.context === 'verse') || (block.type === 'block.end' && block.name.toUpperCase() === 'VERSE')) {
          if (block.name.toUpperCase() === 'VERSE') {
            // verse blocks can only contain objects, so we adjust the
            // lexer context accordingly
            if (block.type === 'block.begin') {
              setContext('verse');
            } else {
              removeContext('verse');
            }
          }
          action();
          return [block];
        }
      }
    }

    if (state.context !== 'verse') {
      const list = tokenizeListItem({ reader })
      if (list.length > 0) return list
    }

    if (l.startsWith('# ') && state.context !== 'verse') {
      const comment = match(/^#\s+(.*)$/)
      if (comment) {
        eat('line')
        return [ {
          type: 'comment',
          position: comment.position,
          value: comment.captures[1],
        } ]
      }
    }

    const drawer = tokenizeDrawer({ reader })
    if (drawer.length > 0) return drawer
    const table = tokenizeTable({ reader })
    if (table.length > 0) return table

    const hr = eat(/^\s*-{5,}\s*$/).position
    if (!isEmpty(hr)) {
      return [{
        type: 'hr',
        position: hr,
      }]
    }

    if (now().column === 1) {
      const footnote = tokenizeFootnote({ reader })
      if (footnote.length > 0) return footnote
    }

    // last resort
    return inlineTok({ reader })
  }

  const peek = (offset = 0) : Token | undefined => {
    const pos = cursor + offset
    if (pos >= tokens.length) {
      tokens = tokens.concat(tok())
    }
    return tokens[pos]
  }

  const modify = (f: (t: Token) => Token, offset = 0): void => {
    const pos = cursor + offset
    const token = peek(offset);
    if (token !== undefined) {
      tokens[pos] = f(token);
    }
  }

  const _eat = (type: string | undefined = undefined): Token | undefined => {
    const t = peek()
    if (!t) return undefined
    if (!type || type === t.type) {
      cursor += 1
      return t
    }
    return undefined
  }

  return {
    peek,
    eat: _eat,
    eatAll: (type: string): number => {
      let count = 0
      while(_eat(type)) { count += 1 }
      return count
    },
    match: (cond, offset = 0) => {
      const token = peek()
      if (!token) return false
      if (typeof cond === 'string') {
        return token.type === cond
      }
      return cond.test(token.type)
    },

    all: (max: number | undefined = undefined) : Token[] => {
      let _all: Token[] = []
      let tokens = tok()
      while (tokens.length > 0) {
        _all = _all.concat(tokens)
        tokens = tok()
      }
      return _all
    },

    save: () => cursor,

    restore: (point) => { cursor = point },

    addInBufferTodoKeywords: (text) => {
      inBufferTodoKeywordSets.push(todoKeywordSet(text))
    },
    substring,
    modify
  };
}
