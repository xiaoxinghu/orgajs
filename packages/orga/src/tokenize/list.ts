import { Token } from '../types'
import { Reader } from '../reader'
import { tokenize as tokenizeInline } from './inline'
import * as tk from './util';

interface Props {
  reader: Reader;
}

export default ({ reader }: Props): Token[] => {
  const { now, match, eat, jump, substring } = reader

  let tokens: Token[] = []

  const indent = now().column - 1

  const bullet = match(/^([-+]|\d+[.)])(?=\s)/)
  if (!bullet) return []
  tokens.push(tk.tokListBullet(indent, /^\d/.test(bullet.captures[1]), { position: bullet.position }));

  jump(bullet.position.end)
  eat('whitespaces')

  const checkbox = match(/^\[(x|X|-| )\](?=\s)/)
  if (checkbox) {
    const position = checkbox.position;
    tokens.push(tk.tokListCheckbox(checkbox.captures[1] !== ' ', { position }));
    jump(position.end)
  }

  eat('whitespaces')

  const tagMark = match(/\s+::\s+/)
  if (tagMark) {
    const position = { start: now(), end: tagMark.position.start }
    tokens.push(tk.tokListItemTag(substring(position), { position }));
    jump(tagMark.position.end)
  }

  tokens = tokens.concat(tokenizeInline({ reader }))

  eat('line')

  return tokens
}
