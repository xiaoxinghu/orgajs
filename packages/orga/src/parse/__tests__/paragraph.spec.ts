import { tokenize } from '../../tokenize'
import { parse } from '../index'

import {
  anonFootnote,
  footnoteReference,
  inlineFootnote,
  paragraph,
  pos,
  testParse,
  text
} from './util';

describe('Parse Paragraph', () => {
  it('works', () => {
    const content = `
#+ATTR_HTML: :style background: black;
[[https://github.com/xiaoxinghu/orgajs][Here's]] to the *crazy* ones, the /misfits/, the _rebels_, the ~troublemakers~,
the round pegs in the +round+ square holes...

#+ATTR_HTML: :width 300
[[file:logo.png]]
`

    //     const content = `
    // [[https://github.com/xiaoxinghu/orgajs][Here's]] some /content/ hello world.
    // `
    expect(parse(tokenize(content))).toMatchSnapshot()
    // debug(content)
  })

  function testParagraph(testName: string, inText: string, ...expected: Parameters<typeof paragraph>) {
    return testParse(testName, inText, [paragraph(...expected)]);
  }

  testParagraph('with standard footnote',
    'hello[fn:named] world.', [
    text('hello'),
    footnoteReference('named'),
    text(' world.')
  ]);

  testParagraph('with inline footnote',
    'hello[fn:named:Inline named footnote] world.', [
    text('hello'),
    inlineFootnote('named', [text('Inline named footnote')]),
    text(' world.')
  ]);

  testParagraph('with anonymous footnote',
    'hello[fn::Anonymous footnote] world.', [
    text('hello'),
    anonFootnote([text('Anonymous footnote')]),
    text(' world.')
  ]);

  testParagraph('with anonymous with no body',
    'hello[fn::] world.', [
    text('hello'),
    anonFootnote([text('', { position: pos([1, 11], [1, 11]) })]),
    text(' world.')
  ]);

  testParagraph('with anonymous nested footnote',
    'hello[fn::An [fn::Anonymous footnote]!] world.', [
    text('hello'),
    anonFootnote([text('An '), anonFootnote([text('Anonymous footnote')]), text('!')]),
    text(' world.')
  ]);
})
