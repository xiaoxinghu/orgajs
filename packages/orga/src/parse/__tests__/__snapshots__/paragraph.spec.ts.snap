// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parse Paragraph works 1`] = `
Object {
  "children": Array [
    Object {
      "attributes": Object {
        "attr_html": Object {
          "style": "background: black;",
        },
      },
      "children": Array [
        Object {
          "description": "Here's",
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 49,
              "line": 3,
              "offset": 88,
            },
            "start": Object {
              "column": 1,
              "line": 3,
              "offset": 40,
            },
          },
          "protocol": "https",
          "search": undefined,
          "type": "link",
          "value": "https://github.com/xiaoxinghu/orgajs",
        },
        Object {
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 57,
              "line": 3,
              "offset": 96,
            },
            "start": Object {
              "column": 49,
              "line": 3,
              "offset": 88,
            },
          },
          "type": "text.plain",
          "value": " to the ",
        },
        Object {
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 64,
              "line": 3,
              "offset": 103,
            },
            "start": Object {
              "column": 57,
              "line": 3,
              "offset": 96,
            },
          },
          "type": "text.bold",
          "value": "crazy",
        },
        Object {
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 75,
              "line": 3,
              "offset": 114,
            },
            "start": Object {
              "column": 64,
              "line": 3,
              "offset": 103,
            },
          },
          "type": "text.plain",
          "value": " ones, the ",
        },
        Object {
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 84,
              "line": 3,
              "offset": 123,
            },
            "start": Object {
              "column": 75,
              "line": 3,
              "offset": 114,
            },
          },
          "type": "text.italic",
          "value": "misfits",
        },
        Object {
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 90,
              "line": 3,
              "offset": 129,
            },
            "start": Object {
              "column": 84,
              "line": 3,
              "offset": 123,
            },
          },
          "type": "text.plain",
          "value": ", the ",
        },
        Object {
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 98,
              "line": 3,
              "offset": 137,
            },
            "start": Object {
              "column": 90,
              "line": 3,
              "offset": 129,
            },
          },
          "type": "text.underline",
          "value": "rebels",
        },
        Object {
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 104,
              "line": 3,
              "offset": 143,
            },
            "start": Object {
              "column": 98,
              "line": 3,
              "offset": 137,
            },
          },
          "type": "text.plain",
          "value": ", the ",
        },
        Object {
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 119,
              "line": 3,
              "offset": 158,
            },
            "start": Object {
              "column": 104,
              "line": 3,
              "offset": 143,
            },
          },
          "type": "text.code",
          "value": "troublemakers",
        },
        Object {
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 120,
              "line": 3,
              "offset": 159,
            },
            "start": Object {
              "column": 119,
              "line": 3,
              "offset": 158,
            },
          },
          "type": "text.plain",
          "value": ",",
        },
        Object {
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 1,
              "line": 4,
              "offset": 160,
            },
            "start": Object {
              "column": 120,
              "line": 3,
              "offset": 159,
            },
          },
          "type": "text.plain",
          "value": " ",
        },
        Object {
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 23,
              "line": 4,
              "offset": 182,
            },
            "start": Object {
              "column": 1,
              "line": 4,
              "offset": 160,
            },
          },
          "type": "text.plain",
          "value": "the round pegs in the ",
        },
        Object {
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 30,
              "line": 4,
              "offset": 189,
            },
            "start": Object {
              "column": 23,
              "line": 4,
              "offset": 182,
            },
          },
          "type": "text.strikeThrough",
          "value": "round",
        },
        Object {
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 46,
              "line": 4,
              "offset": 205,
            },
            "start": Object {
              "column": 30,
              "line": 4,
              "offset": 189,
            },
          },
          "type": "text.plain",
          "value": " square holes...",
        },
      ],
      "parent": [Circular],
      "position": Object {
        "end": Object {
          "column": 1,
          "line": 6,
          "offset": 207,
        },
        "start": Object {
          "column": 1,
          "line": 3,
          "offset": 40,
        },
      },
      "type": "paragraph",
    },
    Object {
      "attributes": Object {
        "attr_html": Object {
          "width": 300,
        },
      },
      "children": Array [
        Object {
          "description": undefined,
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 18,
              "line": 7,
              "offset": 248,
            },
            "start": Object {
              "column": 1,
              "line": 7,
              "offset": 231,
            },
          },
          "protocol": "file",
          "search": undefined,
          "type": "link",
          "value": "logo.png",
        },
      ],
      "parent": [Circular],
      "position": Object {
        "end": Object {
          "column": 19,
          "line": 7,
          "offset": 249,
        },
        "start": Object {
          "column": 1,
          "line": 7,
          "offset": 231,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 19,
      "line": 7,
      "offset": 249,
    },
    "start": Object {
      "column": 1,
      "line": 3,
      "offset": 40,
    },
  },
  "properties": Object {},
  "type": "document",
}
`;
